### DISCUSSION TITLE
How to send a custom SysEx message

### DISCUSSION THREAD

**Discussion by MeestorX** [melanchall/drywetmidi#199] [General]
I have a MIDI SysEx message that doesn't use a length byte. (I've actually never encountered one that did!)
It seems when creating a `NormalSysExEvent`, it doesn't allow me to include the 0xF0 at the beginning (it errors out) and if I create one without the 0xF0 at the beginning, it seems to add this length byte that I don't want.

If I try `NormalSysExEvent HS = new NormalSysExEvent(new byte[] {0x00 0x00 0x66 0x14 0x00 xF7})`
It changes it to `0xF0 0x06 0x00 0x00 0x66 0x14 0xF7`. I don't want it to add the length byte.

How can I send a SysEx message like this:

0xF0 0x00 0x00 0x66 0x14 0x00 xF7

**Comment by melanchall** [comment]
Please include the code you use to "send" a sys ex event and provide more details. If you mean sending to an output device, than I say that DryWetMIDI adds nothing except `0xF0` to event's data. You can easily see this in the library code here:

https://github.com/melanchall/drywetmidi/blob/9382437a94fff4c80a5b9be55b177887200df309/DryWetMidi/Multimedia/OutputDevice/OutputDevice.cs#L517

You can also see my [answer in another issue](https://github.com/melanchall/drywetmidi/issues/172#issuecomment-1092594345) where I show sending/receiving sys ex event with DryWetMIDI via virtual device and all received bytes are exactly the same as sent ones, no length added.

**Comment by MeestorX** [comment]
Ok, thank you for verifying.
Perhaps there's an issue in `MidiEventToBytesConverter()`?
```byte[] HB = new byte[] {0x00, 0x00, 0x66, 0x14, 0x00, 0xF7};
NormalSysExEvent HS = new NormalSysExEvent(HB);
Console.WriteLine(BitConverter.ToString(HS.Data)); // 00-00-66-14-00-F7

byte[] HS_AsBytes = new MidiEventToBytesConverter().Convert(HS)
Console.WriteLine(BitConverter.ToString(HS_AsBytes)); // F0-06-00-00-66-14-00-F7
```

**Comment by melanchall** [comment]
Well, so your issue is with `MidiEventToBytesConverter`, not with "sending". `MidiEventToBytesConverter` converts MIDI data according to Standard MIDI Files (SMF) specification which states:

> \<sysex event\> is used to specify a MIDI system exclusive message, either as one unit or in
packets, or as an "escape" to specify any arbitrary bytes to be transmitted. A normal complete
system exclusive message is stored in a MIDI File in this way:
>
> F0 \<length\> \<bytes to be transmitted after F0\>

So the behavior you see is absolutely correct. Can you please describe in more details why you need to use `MidiEventToBytesConverter` and why you expect it should violate SMF spec?

Anyway you can just combine `0xF0` with `HS.Data` to get what you want.

**Comment by MeestorX** [comment]
Thank you for clarifying. I was simply using `MidiEventToBytesConverter` to verify the exact bytes that are being sent. I use it for debugging all over the place.
Is there a way to display the exact bytes of a MidiEvent if MidiEventToBytesConverter is not it?

**Reply by melanchall** [reply]
I've already provided you the link to the source code for sys ex event sending:

https://github.com/melanchall/drywetmidi/blob/9382437a94fff4c80a5b9be55b177887200df309/DryWetMidi/Multimedia/OutputDevice/OutputDevice.cs#L517

You can see there exact bytes of a sys ex event – status byte (`0xF0`) + sys ex data bytes. I believe it doesn't make sense to "learn" what bytes are sent because you have much more – the source code. More than that with Windows API short events are sent as integer numbers. More that that – in reverse order. So there is no profit in converting events to bytes since events can be sent differently for different types or operating systems.

Can you please explain why you need such strange debugging? Why not just send an event and be happy? :) Because sending is responsibility of the DryWetMIDI, it looks pointless to debug this process unless you see some bugs. If you see bugs in the library, please provide me info about them.

**Comment by MeestorX** [comment]
To clarify, I'm using devices, not files.

**Comment by MeestorX** [comment]
I'm sending/receiving data from TCP, MIDI, Osc and other methods, and my send/receive routine is generic for all types, in that it translates everything to byte[] for easy conversion. I can then verify that things are being sent/received and converted correctly so, it's handy having a method to convert all MIDI to bytes as well to allow comparisons.

**Reply by melanchall** [reply]
Anyway there is no way to see bytes of sys ex events with `MidiEventToBytesConverter` as you want. For `NormalSysExEvent` just concatenate `0xF0` with event data and you'll get what you want.

**Reply by MeestorX** [reply]
Ok. That's too bad, but I can deal with it, now that I know.

Thanks!

**Reply by melanchall** [reply]
@MeestorX Well, I've published prerelease version of the library – [6.1.2-prerelease1](https://www.nuget.org/packages/Melanchall.DryWetMidi/6.1.2-prerelease1). Now you can set `BytesFormat` property on converters to `BytesFormat.Device` which switches how sys ex (and some other events) are read/written.

**Reply by MeestorX** [reply]
Wow! You really work hard on this library. Appreciate all you've done.
So BytesFormat.Device will not add the length byte? Does it still add the 0xF0 at the start?

**Reply by melanchall** [reply]
> So BytesFormat.Device will not add the length byte?

Yes.

> Does it still add the 0xF0 at the start?

Of course.

Thanks for using the library!

**Comment by melanchall** [comment]
:rocket: **6.1.2 version is released now!**

Prerelease NuGet packages will be unlisted soon, so please update the package references to the new version.

Thanks for using the library!
